### PR TITLE
Replace tokio with smol

### PR DESCRIPTION
**Pull request by mrwwheat7**

### CODE REVIEW

File: src/watch/mod.rs [hunk 1]
```diff
@@ -162,6 +162,7 @@ impl Watch {
         tunnel
             .req_sender
             .send(WatchRequest::create(key_range))
+            .await
             .expect("emit watch request");
```
Comment (pwang7) [inline src/watch/mod.rs]: use unwrap_or_else instead of expect

File: src/watch/mod.rs [hunk 2]
```diff
@@ -102,17 +101,18 @@ impl WatchTunnel {
...
                     Ok(Some(resp)) => {
-                        resp_sender.send(Ok(From::from(resp))).unwrap();
+                        resp_sender.send(Ok(From::from(resp))).await.unwrap();
                     }
                     Ok(None) => {
                         return;
                     }
                     Err(e) => {
-                        resp_sender.send(Err(From::from(e))).unwrap();
+                        resp_sender.send(Err(From::from(e))).await.unwrap();
```
Comment (pwang7) [inline src/watch/mod.rs]: use unwrap_or_else instead of unwrap

File: src/lease/mod.rs [hunk 3]
```diff
@@ -209,6 +209,7 @@ impl Lease {
             .await
             .req_sender
             .send(req)
+            .await
             .map_err(|_| Error::ChannelClosed)
```
Comment (pwang7) [inline src/lease/mod.rs:162]: do not use map_err, use this_error to handle errors in a unified way

File: Cargo.toml [hunk 4]
```diff
@@ -1,25 +1,27 @@
 [package]
-name = "etcd-rs"
+name = "etcd-client"
 version = "0.3.0"
```
Comment (pwang7) [inline Cargo.toml]: version change to 0.1.0

File: src/lease/mod.rs [hunk 5]
```diff
@@ -1,75 +1,13 @@
 //! Leases are a mechanism for detecting client liveness. The cluster grants leases with a time-to-live. A lease expires if the etcd cluster does not receive a keepAlive within a given TTL period.
-//!
-//! # Examples
```
Comment (pwang7) [inline src/lease/mod.rs:3]: please keep the example

File: src/lease/mod.rs [hunk 6]
```diff
@@ -122,17 +60,18 @@ impl LeaseKeepAliveTunnel {
                 };
                 match resp {
                     Ok(Some(resp)) => {
-                        resp_sender.send(Ok(From::from(resp))).unwrap();
+                        resp_sender.send(Ok(From::from(resp))).await.unwrap();
```
Comment (pwang7) [inline src/lease/mod.rs]: do not use unwrap, use unwrap_or_else instead

File: src/lease/mod.rs [hunk 7]
```diff
@@ -150,7 +89,7 @@ impl LeaseKeepAliveTunnel {
 impl Shutdown for LeaseKeepAliveTunnel {
     async fn shutdown(&mut self) -> Result<()> {
         if let Some(shutdown) = self.shutdown.take() {
-            shutdown.send(()).map_err(|_| Error::ChannelClosed)?;
+            shutdown.send(()).await.map_err(|_| Error::ChannelClosed)?;
```
Comment (pwang7) [inline src/lease/mod.rs:102]: this kind of error conversion lost the original error context, use this_error to auto convert error by using ?

File: src/lazy.rs [hunk 8]
```diff
@@ -2,7 +2,7 @@
 use std::ops::{Deref, DerefMut};
 
 use async_trait::async_trait;
-use tokio::sync::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+use smol::lock::{RwLock, RwLockReadGuard, RwLockWriteGuard};
 
 use crate::Result;
```
Comment (pwang7) [inline src/lazy.rs:8]: what the `Lazy` structure for?
looks like it can be replaced by `LazyStatic`

**Review by pwang7 — CHANGES_REQUESTED** [review]
enable the link check in lib.rs

File: src/lease/mod.rs (hunk 5 above)
Comment (mrwwheat7) [reply src/lease/mod.rs:3]: The example is written by tokio, I can rewrite them by smol later. WDYT

File: src/lazy.rs (hunk 8 above)
Comment (mrwwheat7) [reply src/lazy.rs:8]: It is similar with LazyStatic, but it also provide Lock related functionality

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -6,21 +6,31 @@ on:
   schedule: [cron: "0 0 * * *"]
 
 env:
-  CARGO_TERM_COLOR: always
+  CI_RUST_TOOLCHAIN: 1.47.0
 
 jobs:
   test:
     name: Rust ${{matrix.rust}}
```
Comment (pwang7) [inline .github/workflows/ci.yml]: name: Test

File: src/lazy.rs (hunk 8 above)
Comment (pwang7) [reply src/lazy.rs:8]: Why it needs lock? Will the lock impact performance?

File: src/lease/mod.rs [hunk 10]
```diff
@@ -122,17 +60,24 @@ impl LeaseKeepAliveTunnel {
...
                     }
                     Ok(None) => {
                         return;
                     }
                     Err(e) => {
-                        resp_sender.send(Err(From::from(e))).unwrap();
+                        resp_sender
+                            .send(Err(From::from(e)))
+                            .await
+                            .unwrap_or_else(|_| panic!());
```
Comment (pwang7) [inline src/lease/mod.rs]: proper panic and error handling!

`unwrap_or_else(|e| panic!("failed to send response, the error is: {}", e));`

File: src/lease/mod.rs [hunk 11]
```diff
@@ -141,7 +86,7 @@ impl LeaseKeepAliveTunnel {
         }
     }
 
-    fn take_resp_receiver(&mut self) -> UnboundedReceiver<Result<LeaseKeepAliveResponse>> {
+    fn take_resp_receiver(&mut self) -> Receiver<Result<LeaseKeepAliveResponse>> {
         self.resp_receiver.take().unwrap()
```
Comment (pwang7) [inline src/lease/mod.rs]: do not use unwrap()

File: src/watch/mod.rs [hunk 12]
```diff
@@ -102,17 +59,24 @@ impl WatchTunnel {
                 };
                 match resp {
                     Ok(Some(resp)) => {
-                        resp_sender.send(Ok(From::from(resp))).unwrap();
+                        resp_sender
+                            .send(Ok(From::from(resp)))
+                            .await
+                            .unwrap_or_else(|_| panic!());
```
Comment (pwang7) [inline src/watch/mod.rs]: proper error and panic handling!

File: src/watch/mod.rs [hunk 13]
```diff
@@ -102,17 +59,24 @@ impl WatchTunnel {
...
                     }
                     Ok(None) => {
                         return;
                     }
                     Err(e) => {
-                        resp_sender.send(Err(From::from(e))).unwrap();
+                        resp_sender
+                            .send(Err(From::from(e)))
+                            .await
+                            .unwrap_or_else(|_| panic!());
```
Comment (pwang7) [inline src/watch/mod.rs]: proper panic and error handling!

File: src/watch/mod.rs [hunk 14]
```diff
@@ -102,17 +59,24 @@ impl WatchTunnel {
...
                     }
                 };
             }
-        });
+        })
+        .detach();
```
Comment (pwang7) [inline src/watch/mod.rs:83]: if detach, you might need to config how many pending tasks can run in parallel

File: src/watch/mod.rs [hunk 15]
```diff
@@ -162,7 +126,8 @@ impl Watch {
         tunnel
             .req_sender
             .send(WatchRequest::create(key_range))
-            .expect("emit watch request");
+            .await
+            .unwrap_or_else(|_| panic!("emit watch request"));
```
Comment (pwang7) [inline src/watch/mod.rs]: `unwrap_or_else` should use the input error

**Review by pwang7 — CHANGES_REQUESTED** [review]

File: src/lease/mod.rs (hunk 3 above)
Comment (mrwwheat7) [reply src/lease/mod.rs:162]: Synced offline. Will address this later.

File: src/lease/mod.rs (hunk 7 above)
Comment (mrwwheat7) [reply src/lease/mod.rs:102]: Synced offline. Will address this later.

File: src/lazy.rs (hunk 8 above)
Comment (mrwwheat7) [reply src/lazy.rs:8]: It is for lease and watch, since lease and watch keep a Tunnel, the lock is to keep the order of Request and Response, it can make sure the keep_alive Request is send before the keep_alive_responses is returned.

And it may impact the performance, since when `keep_alive` method is called, it will acquire write lock, so other method will be blocked.

File: src/watch/mod.rs (hunk 14 above)
Comment (mrwwheat7) [reply src/watch/mod.rs:83]: Yes, but for now, one etcd client will only have one WatchTunnel Task.

File: src/lazy.rs (hunk 8 above)
Comment (pwang7) [reply src/lazy.rs:8]: Add an issue to track this lock, consider refactor it using lock-free data structure.
